tc/autotuner/autotuner.h

```diff
@@ -0,0 +1,229 @@
+  /// across devices for the purpose of running correctness checks during
+  /// tuning (future work).
+  const std::unordered_map<size_t, std::vector<const DLConstTensor*>> kInputs_;
+  std::unordered_map<size_t, std::vector<const DLTensor*>> outputs_;
+
+  // results
+  size_t bestTime_;
+  MappingOptionsType bestMappingOptions_;
+
+  // backing options cache
+  std::shared_ptr<OptionsCache<Backend>> optionsCache_;
+};
+} // namespace detail
+
+/**
+ * An Autotuner provides the basic interface to run a SearchStrategy over a
+ * particular Backend.
+ * Under the hood it instantiates a TuningHarness<Backend> and a
+ * SearchStrategy. The Autotuner then calls TuningHarness<Backend>::run()
+ * on the SearchStrategy.
+ */
+template <typename Backend, typename SearchStrategy>
+class Autotuner {
+ public:
+  using MappingOptionsType = typename Backend::MappingOptionsType;
+  using OptionsCacheType = tc::autotune::OptionsCache<Backend>;
+
+  /// Parse and store the underlying tree representation for all the TC
+  /// functions provided.
+  explicit Autotuner(const std::string& tc);
+
+  /// Runs autotuning on the TC function tcEntryPoint.
+  /// This assumes input and output tensors of proper sizes have been
+  /// initialized on each device. For now this initialization is left to the
+  /// client of the Autotuner.
+  /// Tuning performs a search of template type SearchStrategy starting from
+  /// baseMapping.
+  ///
+  /// Alternatively an OptionsCache cacheFileName serialized path
+  /// can be specified from which the tuner recovers multiple starting points
+  /// and appends to the baseMapping. This can be useful in a reinforcement
+  /// situation where short tunings are run and their results cached
+  /// iteratively.
+  ///
+  /// Lastly a TuningParameterFixer function can be specified to limit the
+  /// search space (i.e. when certain parameters are known to be good/bad
+  /// independently on a particular TC).
+  ///
+  /// \return a vector MappingOptions, if it is empty then tuning did not find
+  /// a single good configuration. This should be a very rare occurrence but
+  /// it is possible in particular if the skipExecutionOrWarmup functin is too
+  /// aggressive and the problem size is too small.
+  std::vector<MappingOptionsType> tune(
+      const std::string& tcEntryPoint,
+      const std::unordered_map<size_t, std::vector<const DLConstTensor*>>&
+          inputs,
+      std::unordered_map<size_t, std::vector<const DLTensor*>>& outputs,
+      const MappingOptionsType& baseMapping,
+      const std::string& cacheFileName = "",
+      const TuningParameterFixer& fixedParams = TuningParameterFixer());
+
+ private:
+  std::map<std::string, lang::TreeRef> tcEntryPointMap_;
+  std::shared_ptr<OptionsCache<Backend>> optionsCache_;
+};
+
+/// Helper functions that need specializing for various backends.
+/// The implementation is not set in stone but just a first approximation to
+/// break the CUDA dependence.
```

Comment:
is this still relevant?

Reply:
unfortunately yes, I still haven't decided what to do with these, they are related to TuningConfiguration which I am leaving for a later time